Replace `NOTIFICATION_DEBUG_COLLISIONS_HINT_CHANGED` with a signal

KoBeWi opened the pull request:
> [!NOTE]
This PR was reworked and some comments are no longer relevant.

Replaces the new notification added in #119816 with a signal. The reason is that notifications are a node-specific construct, which can't be used in physics code that doesn't use nodes. See https://github.com/godotengine/godot-proposals/issues/7208 for a related problem.

The new signal is not exposed right now, but that can be changed.

smix8 commented:
Using a loop over the entire root tree and using hardcoded node class names in string lists on top looks like a hyper-mislead hacky way of doing this.

This should be the PhysicsServer sending out a debug related signal that all the physics related nodes with debug are connected with to act on a debug change.

That would work for both 2D and 3D without issues as can be seen by other debug options outside physics that already do that. It is also the only thing that will work should physics ever manage to display server debug.

carsonetb commented:
May I ask why you created a whole new `_config_capture` function instead of just using the `parse_message` one?

KoBeWi commented:
Because I don't know how to use debugger.

carsonetb commented:
> This should be the PhysicsServer sending out a debug related signal that all the physics related nodes with debug are connected with to act on a debug change.

@smix8 Apparently signals aren't used by the engine, so you'd have to have some kind of workaround for that. (or maybe I'm missing something?)

I'd like to see an example of other debug options doing that outside of physics.

KoBeWi commented:
I added `_debug_options_changed` signal to PhysicsServer2D and it's used instead of hard-coded list.
I also discovered that TileMapLayers weren't updating properly, so I fixed it in very inefficient way.

carsonetb commented:
Looks like a lot of the checks are failing because you have the physics signal trying to connect in [touch_screen_button.cpp](https://github.com/godotengine/godot/pull/102963/files#diff-0f41ceaa3743ed1135c5221a3e370917f7642d7a439a47bc54c582c0864c9767), which is strange.

Also, if it's any help, here's some GDScript code from my suggestion that someone posted for toggling 3D collision shape visibility, not sure if it's applicable here:

```
func toggle_3d_collision_shape_visibility() -> void:
	var tree: SceneTree = get_tree()
	# https://github.com/godotengine/godot-proposals/issues/2072
	tree.debug_collisions_hint = not tree.debug_collisions_hint
	print("Set show_debug_collisions_hint: ", tree.debug_collisions_hint)

	# Traverse tree to call toggle collision visibility
	var node_stack: Array[Node] = [tree.get_root()]
	while not node_stack.is_empty():
		var node: Node = node_stack.pop_back()
		if is_instance_valid(node):
			if node is RayCast3D \
				or node is CollisionShape3D \
				or node is CollisionPolygon3D \
				#or node is CollisionObject3D \ (Doesn't work well with police-mans (they utilize exit_tree()).
				or node is GPUParticlesCollision3D \
				or node is GPUParticlesCollisionBox3D \
				or node is GPUParticlesCollisionHeightField3D \
				or node is GPUParticlesCollisionSDF3D \
				or node is GPUParticlesCollisionSphere3D \
				or node is CSGPrimitive3D:
				# remove and re-add the node to the tree to force a redraw
				# https://github.com/godotengine/godot/blob/26b1fd0d842fa3c2f090ead47e8ea7cd2d6515e1/scene/3d/collision_object_3d.cpp#L39
				var parent: Node = node.get_parent()
				if parent:
					parent.remove_child(node)
					parent.add_child(node)
			node_stack.append_array(node.get_children())
```

KoBeWi commented:
Your solution involves removing and adding nodes to scene tree, which might have unwanted side-effects.

carsonetb commented:
Yeah that's true. But it does at least show that the setting for this does update at run-time, which is weird because you might add a collision shape in and see debug for it but not others.

AThousandShips commented on scene/debugger/scene_debugger.cpp:
```diff
@@ -141,6 +142,28 @@ void SceneDebugger::_handle_embed_input(const Ref<InputEvent> &p_event, const Di
+	if (p_msg == "debug_collisions_hint") {
+		scene_tree->set_debug_collisions_hint(p_args[0]);
+		PhysicsServer2D::get_singleton()->emit_signal("_debug_options_changed");
+		// PhysicsServer3D::get_singleton()->emit_signal("_debug_options_changed"); // TODO
```
```suggestion
		//PhysicsServer3D::get_singleton()->emit_signal("_debug_options_changed"); // TODO
```
No space for commented out code

KoBeWi committed:
Replace NOTIFICATION_DEBUG_COLLISIONS_HINT_CHANGED with a signal

KoBeWi commented:
Rebased after #119816
The PR was reworked, see OP for details.

KoBeWi commented:
After a discussion on chat, I reworked my rework.

Physics servers are now responsible for keeping track of debug hint, which was renamed to simply `debug_enabled`. There are new related methods and a signal in each server. The SceneTree's debug hint is effectively deprecated, but I didn't deprecate it properly yet, because the new methods are not exposed right now.
